File: xamarin/Xtrade/Xtrade.AndroidApp/Activities/HomeActivity.cs
namespace Xtrade.AndroidApp.Activities
{
    using System;
    using Android.App;
    using Android.OS;
    using Android.Support.Design.Widget;
    using Android.Support.V4.View;
    using Android.Support.V4.Widget;
    using Android.Views;
    using Android.Widget;
    using Fragments;
    using FragmentTransaction = Android.Support.V4.App.FragmentTransaction;
    using Toolbar = Android.Support.V7.Widget.Toolbar;

    [Activity(Label = "@string/applicationName", Theme = "@style/XtradeHome", MainLauncher = true, Icon = "@mipmap/ic_launcher")]
    public class HomeActivity : BaseActivity
    {
        private Toolbar _applicationToolbar;
        private DrawerLayout _drawerLayout;
        private NavigationView _navigationView;

        protected override void OnCreate(Bundle bundle)
        {
            base.OnCreate(bundle);

            this.SetContentView(Resource.Layout.activity_home);

            this._applicationToolbar = this.FindViewById<Toolbar>(Resource.Id.applicationToolbar);
            this._drawerLayout = this.FindViewById<DrawerLayout>(Resource.Id.drawerLayout);
            this._navigationView = this.FindViewById<NavigationView>(Resource.Id.navigationView);

            this.SetSupportActionBar(this._applicationToolbar);
            this.SupportActionBar.SetHomeAsUpIndicator(Resource.Drawable.ic_navigation_menu);
            this.SupportActionBar.SetDisplayHomeAsUpEnabled(true);

            this._navigationView.NavigationItemSelected += this.NavigationViewOnNavigationItemSelected;

            this.ShowAllExchangeRatesFragment();
        }

        public void SetActionBarTitle(string title)
        {
            this.SupportActionBar.Title = title;
        }

        private void NavigationViewOnNavigationItemSelected(object sender, NavigationView.NavigationItemSelectedEventArgs navigationItemSelectedEventArgs)
        {
            navigationItemSelectedEventArgs.MenuItem.SetChecked(true);
            switch (navigationItemSelectedEventArgs.MenuItem.ItemId)
            {
                case Resource.Id.navigation_all_exchange_rates:
                    this._navigationView.SetCheckedItem(Resource.Id.navigation_all_exchange_rates);
                    this.ShowAllExchangeRatesFragment();
                    break;
                case Resource.Id.navigation_calculate:
                    this._navigationView.SetCheckedItem(Resource.Id.navigation_calculate);
                    this.ShowCalculateFragment();
                    break;
            }

            this._drawerLayout.CloseDrawer(GravityCompat.Start);
        }

        public override bool OnOptionsItemSelected(IMenuItem item)
        {
            switch (item.ItemId)
            {
                case Android.Resource.Id.Home:
                    this._drawerLayout.OpenDrawer(GravityCompat.Start);
                    return true;
                default:
                    return base.OnOptionsItemSelected(item);
            }
        }

        public override void OnBackPressed()
        {
            if (this._drawerLayout.IsDrawerOpen(GravityCompat.Start))
            {
                this._drawerLayout.CloseDrawer(GravityCompat.Start);
            }
            else
            {
                base.OnBackPressed();
            }
        }
        
        private void ShowAllExchangeRatesFragment()
        {
            this.ClearFragmentBackStack();

            AllExchangeRatesFragment allExchangeRatesFragment = new AllExchangeRatesFragment();
            FragmentTransaction fragmentTransaction = this.SupportFragmentManager.BeginTransaction();
            fragmentTransaction.Replace(Resource.Id.containerFrameLayout, allExchangeRatesFragment, allExchangeRatesFragment.GetType().Name);
            fragmentTransaction.Commit();
        }

        private void ShowCalculateFragment()
        {
            this.ClearFragmentBackStack();

            CalculateFragment calculateFragment = new CalculateFragment();
            FragmentTransaction fragmentTransaction = this.SupportFragmentManager.BeginTransaction();
            fragmentTransaction.Replace(Resource.Id.containerFrameLayout, calculateFragment, calculateFragment.GetType().Name);
            fragmentTransaction.AddToBackStack(calculateFragment.GetType().Name);
            fragmentTransaction.Commit();
        }

        private void ClearFragmentBackStack()
        {
            for (int i = 0;i < this.SupportFragmentManager.BackStackEntryCount;i++)
            {
                this.SupportFragmentManager.PopBackStackImmediate();
            }
        }
    }
}

File: android/Xtrade/app/src/main/java/nz/co/fendustries/xtrade/activities/HomeActivity.java
package nz.co.fendustries.xtrade.activities;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.NavigationView;
import android.support.v4.app.FragmentTransaction;
import android.support.v4.view.GravityCompat;
import android.support.v4.widget.DrawerLayout;
import android.support.v7.widget.Toolbar;
import android.view.MenuItem;

import nz.co.fendustries.xtrade.R;
import nz.co.fendustries.xtrade.fragments.AllExchangeRatesFragment;
import nz.co.fendustries.xtrade.fragments.CalculateFragment;

/**
 * Created by joshuafenemore on 21/10/16.
 */
public class HomeActivity extends BaseActivity
{
    private Toolbar applicationToolbar;
    private DrawerLayout drawerLayout;
    private NavigationView navigationView;

    @Override
    protected void onCreate(Bundle bundle)
    {
        super.onCreate(bundle);

        this.setContentView(R.layout.activity_home);

        this.applicationToolbar = (Toolbar) this.findViewById(R.id.applicationToolbar);
        this.drawerLayout = (DrawerLayout) this.findViewById(R.id.drawerLayout);
        this.navigationView = (NavigationView) this.findViewById(R.id.navigationView);

        this.setSupportActionBar(this.applicationToolbar);
        this.getSupportActionBar().setHomeAsUpIndicator(R.drawable.ic_navigation_menu);
        this.getSupportActionBar().setDisplayHomeAsUpEnabled(true);

        this.navigationView.setNavigationItemSelectedListener(new NavigationView.OnNavigationItemSelectedListener()
        {
            @Override
            public boolean onNavigationItemSelected(@NonNull MenuItem item)
            {
                item.setChecked(true);
                switch (item.getItemId())
                {
                    case R.id.navigation_all_exchange_rates:
                        showAllExchangeRatesFragment();
                        break;
                    case R.id.navigation_calculate:
                        showCalculateFragment();
                        break;
                }

                drawerLayout.closeDrawer(GravityCompat.START);
                return true;
            }
        });

        if (bundle == null)
        {
            this.showAllExchangeRatesFragment();
        }
    }

    public void setActionBarTitle(String title)
    {
        this.getSupportActionBar().setTitle(title);
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item)
    {
        switch (item.getItemId())
        {
            case android.R.id.home:
                this.drawerLayout.openDrawer(GravityCompat.START);
                return true;
            default:
                return super.onOptionsItemSelected(item);
        }
    }

    @Override
    public void onBackPressed()
    {
        if (this.drawerLayout.isDrawerOpen(GravityCompat.START))
        {
            this.drawerLayout.closeDrawer(GravityCompat.START);
        }
        else
        {
            if (this.getSupportFragmentManager().getBackStackEntryCount() > 0)
            {
                this.getSupportFragmentManager().popBackStackImmediate();
                this.navigationView.setCheckedItem(R.id.navigation_all_exchange_rates);
            }
            else
            {
                super.onBackPressed();
            }
        }
    }

    private void showAllExchangeRatesFragment()
    {
        this.clearFragmentBackStack();

        this.navigationView.setCheckedItem(R.id.navigation_all_exchange_rates);
        AllExchangeRatesFragment allExchangeRatesFragment = new AllExchangeRatesFragment();
        FragmentTransaction fragmentTransaction = this.getSupportFragmentManager().beginTransaction();
        fragmentTransaction.replace(R.id.containerFrameLayout, allExchangeRatesFragment, allExchangeRatesFragment.getClass().getSimpleName());
        fragmentTransaction.commit();
    }

    private void showCalculateFragment()
    {
        this.clearFragmentBackStack();

        this.navigationView.setCheckedItem(R.id.navigation_calculate);
        CalculateFragment calculateFragment = new CalculateFragment();
        FragmentTransaction fragmentTransaction = this.getSupportFragmentManager().beginTransaction();
        fragmentTransaction.replace(R.id.containerFrameLayout, calculateFragment, calculateFragment.getClass().getSimpleName());
        fragmentTransaction.addToBackStack(calculateFragment.getClass().getSimpleName());
        fragmentTransaction.commit();
    }

    private void clearFragmentBackStack()
    {
        for (int i = 0;i < this.getSupportFragmentManager().getBackStackEntryCount();i++)
        {
            this.getSupportFragmentManager().popBackStackImmediate();
        }
    }
}